packages/Core/iterable.candy

```diff
@@ -0,0 +1,167 @@
+bool = use "..bool"
+equals = (use "..equality").equals
+function = use "..function"
+int = use "..int"
+result = use "..result"
+list = use "..list"
+
+# An iterable is a `() => Result[(Item, Iterable[Item]), Any]`.
+# isWithItemType value checkItem :=
+#   result.is value | bool.lazyAnd {
+#     value
+#     | result.map { current ->
+#       list.is current
+#       | bool.lazyAnd { equals (current | list.length) 2 }
+#       | bool.lazyAnd { checkItem (current | list.get 0) }
+#       | bool.lazyAnd (function.is0 (current | list.get 1))
+#     }
+#     | result.unwrapOr True
+#   }
+# is value := isWithItemType value { _ -> True }
+is value := function.is0 value
+
+next iterable :=
+  needs (is iterable)
+  iterable | function.run | result.map { it -> it | list.get 0 }
+isEmpty iterable :=
+  needs (is iterable)
+  iterable | next | result.isOk
+
+## Mapping Transformation
+##
+## These return an iterable with the same length, but different items.
+
+rawMapWithStateHelper = { rawMapWithStateHelper state mapper ->
+  {
+    result = mapper state
+    newItem = result | list.get 0
+    newState = result | list.get 1
+    newItem | result.map { newItem ->
+      (newItem, rawMapWithStateHelper newState mapper)
+    }
+  }
+}
+rawMapWithState initialState mapper :=
+  # `mapper` receives the previous state and returns the optional item to emit
+  # (or none if it's done) and the new state.
+  needs (function.is1 mapper)
+  rawMapWithStateHelper rawMapWithStateHelper initialState mapper
```

Comment:
I'd drop the `raw` prefix or make this non-public or both.

Reply:
Also, this doesn't really map iterables anything, right? For example, our implementation of a list-to-iterable or range-to-iterable converters could also use this. So, maybe call this `iterableWithState` or `iterableFromState` instead to highlight it creates a new iterable from a state and mapper?